zebra-state/src/service.rs

```diff
@@ -1063,6 +1072,32 @@ impl Service<Request> for StateService {
                 .boxed()
             }
 
+            // Used by sync, inbound, and block verifier to check if a block is already in the state
+            // before downloading or validating it.
+            Request::KnownBlock(hash) => {
+                let timer = CodeTimer::start();
+
+                let is_block_queued = self.is_block_queued(&hash);
+
+                let read_service = self.read_service.clone();
+
+                async move {
+                    let response = is_block_queued.then_some(KnownBlock::Queue).or_else(|| {
+                        read::contains(
+                            &read_service.latest_non_finalized_state(),
+                            &read_service.db,
+                            hash,
+                        )
+                    });
```

Comment:
I think that we might accidentally label some non-finalized blocks as `Queued`, because we don't prune `sent_blocks` above the finalized tip.

So I suggest we check for blocks in this order:
* non-finalized state
* queued
* finalized state

This can become a TODO and a doc update to `KnownBlocks` if you don't want to do it now. (The current code doesn't care where the block is.)

Reply:
Thank you!! That's a great idea. Applied in [`7f7d070`](https://github.com/ZcashFoundation/zebra/pull/6335/commits/7f7d07094ab731d18ce3c3a799ec46beda848add).